### PR TITLE
15234 take2 review2

### PR DESCRIPTION
**Pull request by ekaterinadimitrova2**

### CODE REVIEW

File: NEWS.txt [hunk 1]
```diff
@@ -33,6 +33,18 @@ restore snapshots created with the previous major version using the
...
+
+New features
+------------
+
+Upgrading
+---------
+
+Deprecation
+-----------
+
```
Comment (michaelsembwever) [inline NEWS.txt:47]: nit: remove. it'll only confuse the next person…

File: test/conf/cassandra_encryption.yaml [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+  key_alias: testing:1
+  # CBC requires iv length to be 16 bytes
+  # iv_length: 16
+  key_provider:
+    - class_name: org.apache.cassandra.security.JKSKeyProvider
+      parameters:
+        - keystore: test/conf/cassandra.keystore
+          keystore_password: cassandra
+          store_type: JCEKS
+          key_password: cassandra
```
Comment (michaelsembwever) [inline test/conf/cassandra_encryption.yaml:14]: anything change here?

File: src/java/org/apache/cassandra/db/virtual/SettingsTable.java [hunk 3]
```diff
@@ -49,6 +49,13 @@
               .filter(f -> !Modifier.isStatic(f.getModifiers()))
               .collect(Collectors.toMap(Field::getName, Functions.identity()));
 
+    @VisibleForTesting
+    static final Map<String, Field> ANNOTATED_FIELDS =
+    Arrays.stream(Config.class.getFields())
```
Comment (michaelsembwever) [inline src/java/org/apache/cassandra/db/virtual/SettingsTable.java:54]: nit: wrapping is a bit off here. can you wrap with the operator, and indent a bit more , pls.

File: src/java/org/apache/cassandra/config/DataStorageSpec.java [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        };
+
+        /**
+         * Scale d by m, checking for overflow. This has a short name to make above code more readable.
+         */
+        static long x(long d, long m, long over)
+        {
+            if (d > over)
+                return Long.MAX_VALUE;
+            return d * m;
```
Comment (michaelsembwever) [inline src/java/org/apache/cassandra/config/DataStorageSpec.java:338]:
```suggestion
            return Math.multiplyExact(d, m);
```
(to catch programmatic errors where `over` was not correctly defined.

File: src/java/org/apache/cassandra/config/DataRateSpec.java [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        };
+
+        static final double MAX = Long.MAX_VALUE;
+        static final double MEGABITS_PER_MEBIBYTE = 8.388608;
+
+        /**
+         * Scale d by m, checking for overflow. This has a short name to make above code more readable.
+         */
+        static double x(double d, double m, double over)
+        {
```
Comment (michaelsembwever) [inline src/java/org/apache/cassandra/config/DataRateSpec.java:321]: trivial: some way to add an `assert …` here to catch when over is not properly defined to prevent an overflow/POSITIVE_INFINITY ?

File: CHANGES.txt [hunk 6]
```diff
@@ -14,7 +15,6 @@
  * Harden resource management on SSTable components to prevent future leaks (CASSANDRA-17174)
  * Make nodes more resilient to local unrelated files during startup (CASSANDRA-17082)
  * repair prepare message would produce a wrong error message if network timeout happened rather than reply wait timeout (CASSANDRA-16992)
- * Log queries that fail on timeout or unavailable errors up to once per minute by default (CASSANDRA-17159)
```
Comment (michaelsembwever) [inline CHANGES.txt:17]: why is this line removed?

File: conf/cassandra.yaml [hunk 7]
```diff
@@ -394,7 +394,7 @@ key_cache_save_period: 14400
 # headroom for OS block level cache. Do never allow your system to swap.
 #
 # Default value is 0, to disable row caching.
-row_cache_size_in_mb: 0
+row_cache_size: 0MiB
```
Comment (michaelsembwever) [inline conf/cassandra.yaml:397]: does zero need a unit?

File: conf/cassandra.yaml [hunk 8]
```diff
@@ -359,7 +359,7 @@ prepared_statements_cache_size_mb:
 # NOTE: if you reduce the size, you may not get you hottest keys loaded on startup.
 #
 # Default value is empty to make it "auto" (min(5% of Heap (in MB), 100MB)). Set to 0 to disable key cache.
-key_cache_size_in_mb:
+key_cache_size:
```
Comment (michaelsembwever) [inline conf/cassandra.yaml:362]: why do we leave valueless settings uncommented?

File: conf/cassandra.yaml (hunk 7 above)
Comment (ekaterinadimitrova2) [reply conf/cassandra.yaml:397]: Every value comes with a unit

File: conf/cassandra.yaml (hunk 8 above)
Comment (ekaterinadimitrova2) [reply conf/cassandra.yaml:362]: From the comments: # Default value is empty to make it "auto" (min(5% of Heap (in MB), 100MB)). Set to 0 to disable key cache.

File: CHANGES.txt (hunk 6 above)
Comment (ekaterinadimitrova2) [reply CHANGES.txt:17]: squash issue probably, thanks for pointing,  I will return it!

File: conf/cassandra.yaml [hunk 9]
```diff
@@ -1391,35 +1393,35 @@ replica_filtering_protection:
...
-# GC Pauses greater than gc_warn_threshold_in_ms will be logged at WARN level
+# GC Pauses greater than gc_warn_threshold will be logged at WARN level
 # Adjust the threshold based on your application throughput requirement. Setting to 0
 # will deactivate the feature.
-# gc_warn_threshold_in_ms: 1000
+# gc_warn_threshold: 1000ms
 
 # Maximum size of any value in SSTables. Safety measure to detect SSTable corruption
 # early. Any value size larger than this threshold will result into marking an SSTable
 # as corrupted. This should be positive and less than 2048.
```
Comment (michaelsembwever) [inline conf/cassandra.yaml:1423]:
```suggestion
# as corrupted. This should be positive and less than 2GiB.
```

File: src/java/org/apache/cassandra/config/DataRateSpec.java (hunk 5 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: over is used three times with specific constants and there is no expectation to variate so I don't think we really need it

File: src/java/org/apache/cassandra/config/DataStorageSpec.java (hunk 4 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataStorageSpec.java:338]: I was about to commit this but over is used three times with specific constant, I don't think we need this.

File: test/conf/cassandra_encryption.yaml (hunk 2 above)
Comment (ekaterinadimitrova2) [reply test/conf/cassandra_encryption.yaml:14]: Fixed indent while updating all config files and cleaning old issues

File: conf/cassandra.yaml (hunk 8 above)
Comment (michaelsembwever) [reply conf/cassandra.yaml:362]: Sure, but is an empty value the same as a commented out setting? (look at `repair_session_space` which is commented out but also has a calculated default value…)

File: src/java/org/apache/cassandra/config/DataRateSpec.java (hunk 5 above)
Comment (michaelsembwever) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: was just a suggestion. it's defensive programming, who knows what the next coder is thinking… (the method is open for use, and extension for further DataRateUnit classes is feasible…)

File: conf/cassandra.yaml (hunk 8 above)
Comment (ekaterinadimitrova2) [reply conf/cassandra.yaml:362]: Fair point, I will make them consistent. I just verified that by default in config it is null so it won't change any behavior. Thank you for pointing it

Comment (ekaterinadimitrova2) [reply conf/cassandra.yaml:362]: That made me think and I found that the same reasoning applies to `index_summary_capacity`

File: src/java/org/apache/cassandra/config/DataRateSpec.java (hunk 5 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: Probably something like that should be reasonable to keep us in the right boundaries - `assert over < (MAX - 1) && (over == MAX / m);`

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: we can probably also check for being positive, something like that

Comment (michaelsembwever) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: thanks.

what came to mind as an example was CASSANDRA-3492

this was one of my first encounters with the Cassandra dev community.
a simple bug where units were not translated correctly lead to a downed production cluster (worse than that, every node has "corrupted" sstables and could not be started).

the consequences of an overflow or POSITIVE_INFINITY could be as severe.

File: src/java/org/apache/cassandra/config/DataStorageSpec.java (hunk 4 above)
Comment (michaelsembwever) [reply src/java/org/apache/cassandra/config/DataStorageSpec.java:338]: same as https://github.com/apache/cassandra/pull/1425#discussion_r793323808

File: src/java/org/apache/cassandra/db/virtual/SettingsTable.java (hunk 3 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/db/virtual/SettingsTable.java:54]: It was following the same format as FIELDS and overrides but I agree what you said looks better. Should I reformat the other 2? Asking as I know you don't like when we do unrelated changes but I think it will be nice for consistency

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/db/virtual/SettingsTable.java:54]: For the record, I changed this one locally to:
```
    @VisibleForTesting
    static final Map<String, Field> ANNOTATED_FIELDS = Arrays.stream(Config.class.getFields())
                                                             .filter(f -> !Modifier.isStatic(f.getModifiers()))
                                                             .filter(f -> f.isAnnotationPresent(Replaces.class))
                                                             .collect(Collectors.toMap(Field::getName, Functions.identity()));
```

Comment (michaelsembwever) [reply src/java/org/apache/cassandra/db/virtual/SettingsTable.java:54]: Right. If the style has precedence in this file then please leave it.

**Review by michaelsembwever — APPROVED** [review]

File: src/java/org/apache/cassandra/config/DataRateSpec.java (hunk 5 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataRateSpec.java:321]: Agree, thank you for being patient and for the example

File: src/java/org/apache/cassandra/config/DataRateSpec.java [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+         * @return the rate unit corresponding to the given symbol
+         */
+        public static DataRateUnit fromSymbol(String symbol)
+        {
+            for (DataRateUnit value : values())
+            {
+                if (value.symbol.equalsIgnoreCase(symbol))
+                    return value;
+            }
+            throw new IllegalArgumentException(String.format("Unsupported data rate unit: %s. Supported units are: %s",
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/DataRateSpec.java:338]: @michaelsembwever I will change this exception also to ConfigurationException on commit, I missed it

File: test/unit/org/apache/cassandra/config/DataRateSpecTest.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        assertEquals(Integer.MAX_VALUE, new DataRateSpec("9223372036854775807KiB/s").toKibibytesPerSecondAsInt());
+        assertEquals(Long.MAX_VALUE, new DataRateSpec("9223372036854775807MiB/s").toKibibytesPerSecond(), 0);
+        assertEquals(Integer.MAX_VALUE, new DataRateSpec("9223372036854775807MiB/s").toKibibytesPerSecondAsInt());
+
+        assertEquals(Long.MAX_VALUE, new DataRateSpec("9223372036854775807MiB/s").toMebibytesPerSecond(), 0);
+        assertEquals(Integer.MAX_VALUE, new DataRateSpec("9223372036854775807MiB/s").toMebibytesPerSecondAsInt());
+    }
+
+    @Test
+    public void testFromSymbol()
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/config/DataRateSpecTest.java:69]: @michaelsembwever I will change this method on commit, it should be:
```
@Test
    public void testFromSymbol()
    {
        assertEquals(DataRateSpec.DataRateUnit.fromSymbol("B/s"), DataRateSpec.DataRateUnit.BYTES_PER_SECOND);
        assertEquals(DataRateSpec.DataRateUnit.fromSymbol("KiB/s"), DataRateSpec.DataRateUnit.KIBIBYTES_PER_SECOND);
        assertEquals(DataRateSpec.DataRateUnit.fromSymbol("MiB/s"), DataRateSpec.DataRateUnit.MEBIBYTES_PER_SECOND);
        assertThatThrownBy(() -> DataRateSpec.DataRateUnit.fromSymbol("n"))
        .isInstanceOf(ConfigurationException.class)
        .hasMessageContaining("Unsupported data rate unit: n");
    } 
```

File: src/java/org/apache/cassandra/config/DataRateSpec.java [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+    /**
+     * @return the data rate in kibibyts per seconds
+     */
+    public double toKibibytesPerSecond()
+    {
+        return unit.toKibibytesPerSecond(quantity);
+    }
+
+    /**
+     * Returns the data rate in kibibytes per seconds as an {@code int}
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/DataRateSpec.java:152]: per second, not seconds, corrected it at a few places

File: src/java/org/apache/cassandra/config/DataStorageSpec.java (hunk 4 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/DataStorageSpec.java:338]: Two of the unit tests fail (eq and eqHash), trying to get some sense of why now.....

File: src/java/org/apache/cassandra/config/YamlConfigurationLoader.java [hunk 13]
```diff
@@ -137,9 +142,48 @@ public Config loadConfig(URL url) throws ConfigurationException
         }
         catch (YAMLException e)
         {
-            throw new ConfigurationException("Invalid yaml: " + url + SystemUtils.LINE_SEPARATOR
-                                             +  " Error: " + e.getMessage(), false);
+            throw new ConfigurationException("Invalid yaml: " + url, e);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/YamlConfigurationLoader.java:145]: I don't see how this change was intentional, I will return it pre-commit

File: src/java/org/apache/cassandra/config/Config.java [hunk 14]
```diff
@@ -261,96 +300,119 @@
...
 
     public EncryptionOptions.ServerEncryptionOptions server_encryption_options = new EncryptionOptions.ServerEncryptionOptions();
     public EncryptionOptions client_encryption_options = new EncryptionOptions();
 
     public InternodeCompression internode_compression = InternodeCompression.none;
 
-    public int hinted_handoff_throttle_in_kb = 1024;
-    public int batchlog_replay_throttle_in_kb = 1024;
+    @Replaces(oldName = "hinted_handoff_throttle_in_kb", converter = Converters.KIBIBYTES_DATASTORAGE, deprecated = true)
+    public volatile DataStorageSpec hinted_handoff_throttle = new DataStorageSpec("1024KiB");
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/Config.java:358]: If I am not missing something those should stay not volatile. I changed pre-commit

File: src/java/org/apache/cassandra/config/Config.java [hunk 15]
```diff
@@ -261,96 +300,119 @@
...
+    public DataStorageSpec key_cache_size = null;
+    @Replaces(oldName = "key_cache_save_period", converter = Converters.SECONDS_CUSTOM_DURATION, deprecated = true)
+    public volatile DurationSpec key_cache_save_period = new DurationSpec("4h");
     public volatile int key_cache_keys_to_save = Integer.MAX_VALUE;
 
     public String row_cache_class_name = "org.apache.cassandra.cache.OHCProvider";
-    public long row_cache_size_in_mb = 0;
-    public volatile int row_cache_save_period = 0;
+    @Replaces(oldName = "row_cache_size_in_mb", converter = Converters.MEBIBYTES_DATA_STORAGE, deprecated = true)
+    public volatile DataStorageSpec row_cache_size = new DataStorageSpec("0MiB");
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/Config.java:386]: no need of volatile again

File: src/java/org/apache/cassandra/config/Config.java [hunk 16]
```diff
@@ -376,15 +438,21 @@
 
     public final ReplicaFilteringProtectionOptions replica_filtering_protection = new ReplicaFilteringProtectionOptions();
 
-    public volatile Long index_summary_capacity_in_mb;
-    public volatile int index_summary_resize_interval_in_minutes = 60;
+    @Replaces(oldName = "index_summary_capacity_in_mb", converter = Converters.MEBIBYTES_DATA_STORAGE, deprecated = true)
+    public DataStorageSpec index_summary_capacity;
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/Config.java:442]: and here volatile was needed. Adding pre-commit

File: src/java/org/apache/cassandra/config/DurationSpec.java [hunk 17]
```diff
@@ -53,86 +57,112 @@ public Duration(String value)
...
+     * @return a duration
+     */
+    public static DurationSpec inSecondsString(String value)
+    {
+        //parse the string field value
+        Matcher matcher = VALUES_PATTERN.matcher(value);
+
+        long seconds;
+        //if the provided string value is just a number, then we create a Duration Spec value in seconds
+        if (matcher.find())
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/DurationSpec.java:151]: wrong function, we should use, I changed it pre-commit: 
      ```
if (matcher.matches())
         {
             seconds = Long.parseLong(value);
             return new DurationSpec(seconds, TimeUnit.SECONDS);
         }
```

File: src/java/org/apache/cassandra/config/YamlConfigurationLoader.java (hunk 13 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/config/YamlConfigurationLoader.java:145]: Actually we absolutely need this change, otherwise the users won't see the exceptions for units, etc

**Comment by ekaterinadimitrova2** [comment]
@michaelsembwever  This work was squashed and committed almost two weeks ago, should I just close this pull request? I am not sure what is the process as normally we use pull requests against our own forks during review.
